### PR TITLE
expose and set lora_dropout = 0.0

### PR DESCRIPTION
**Pull request by felipemello1**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (please add here)

When using activation offloading, dropout creates a new copy of the tensor, increasing memory usage. The default was 0.05, as recommended by the qlora paper. Here we set it to 0.0, since there is no significant change in eval.

#### Changelog
- exposed lora_dropout in the builders
- Added it to configs
- set dropout = identity if dropout is 0 (unclear if this is needed, but there is no harm)
- missing DoRA docstrings

#### Test plan
run 1 epoch on alpaca with dropout = 0.0 and dropout = 0.05 and eval

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: torchtune/models/llama3_1/_model_builders.py [hunk 1]
```diff
@@ -3,16 +3,13 @@
...
-from typing import List, Optional
+from typing import List
 from functools import partial
 
 from torchtune.models.llama3_1._component_builders import llama3_1, lora_llama3_1
 
 from torchtune.modules import TransformerDecoder
-from torchtune.models.llama3._tokenizer import Llama3Tokenizer
 from torchtune.modules.peft import LORA_ATTN_MODULES
-from torchtune.modules.tokenizers import parse_hf_tokenizer_json
```
Comment (felipemello1) [inline torchtune/models/llama3_1/_model_builders.py:14]: removed those because they werent being used

File: torchtune/models/llama2/_model_builders.py [hunk 2]
```diff
@@ -92,8 +92,7 @@ def lora_llama2_7b(
         use_dora (bool): Decompose the LoRA weight into magnitude and direction, as
             introduced in "DoRA: Weight-Decomposed Low-Rank Adaptation" (https://arxiv.org/abs/2402.09353).
         quantize_base (bool): Whether to quantize base model weights
-        lora_dropout (float): dropout probability for LoRA linear layers. Default: 0.05
```
Comment (RdoubleA) [inline torchtune/models/llama2/_model_builders.py:95]: this got removed

File: torchtune/models/llama3/_model_builders.py [hunk 3]
```diff
@@ -115,6 +115,7 @@ def lora_llama3_8b(
             Default: False
         lora_rank (int): rank of each low-rank approximation
         lora_alpha (float): scaling factor for the low-rank approximation
+        lora_dropout (float): dropout probability for the low-rank approximation. Default: 0.0
```
Comment (RdoubleA) [inline torchtune/models/llama3/_model_builders.py:118]: this wasn't added to the function signature?

File: torchtune/models/llama3_1/_model_builders.py [hunk 4]
```diff
@@ -160,6 +162,9 @@ def lora_llama3_1_70b(
             Default: False
         lora_rank (int): rank of each low-rank approximation
         lora_alpha (float): scaling factor for the low-rank approximation
+        lora_dropout (float): dropout probability for the low-rank approximation
+        use_dora (bool): Decompose the LoRA weight into magnitude and direction, as
```
Comment (RdoubleA) [inline torchtune/models/llama3_1/_model_builders.py:166]: I'm surprised the doclint didn't catch this

File: torchtune/models/qwen2/_model_builders.py [hunk 5]
```diff
@@ -155,6 +155,7 @@ def lora_qwen2_7b(
             Default: False
         lora_rank (int): rank of each low-rank approximation
         lora_alpha (float): scaling factor for the low-rank approximation
+        lora_droppout (float): dropout probability for the low-rank approximation. Default: 0.0
```
Comment (RdoubleA) [inline torchtune/models/qwen2/_model_builders.py:158]:
```suggestion
        lora_dropout (float): dropout probability for the low-rank approximation. Default: 0.0
```

File: torchtune/models/qwen2/_model_builders.py [hunk 6]
```diff
@@ -206,6 +207,7 @@ def lora_qwen2_0_5b(
...
```
Comment (RdoubleA) [inline torchtune/models/qwen2/_model_builders.py:210]:
```suggestion
        lora_dropout (float): dropout probability for the low-rank approximation. Default: 0.0
```

File: torchtune/modules/peft/lora.py [hunk 7]
```diff
@@ -65,7 +65,7 @@ def __init__(
         self.register_parameter(
             "bias", nn.Parameter(bias) if bias is not None else None
         )
-        self.dropout = nn.Dropout(p=dropout)
+        self.dropout = nn.Dropout(p=dropout) if dropout > 0.0 else nn.Identity()
```
Comment (joecummings) [inline torchtune/modules/peft/lora.py:68]: Did you try inplace or no?

Comment (felipemello1) [reply torchtune/modules/peft/lora.py:68]: i didnt, but i googled a bit about it and saw that people reporting issues with backward computation. In my experience, pytorch doesnt like inplace computations, because it needs the input preserved for gradient calculation.

**Review by ebsmothers — APPROVED** [review]
Looks great!
